### PR TITLE
feat: add `allowEmptyCase` option to no-fallthrough rule

### PR DESCRIPTION
**Pull request by amareshsm**
#### Prerequisites checklist

- [x] I have read the [contributing guidelines](https://github.com/eslint/eslint/blob/HEAD/CONTRIBUTING.md).

#### What is the purpose of this pull request? (put an "X" next to an item)

[ ] Documentation update
[ ] Bug fix ([template](https://raw.githubusercontent.com/eslint/eslint/HEAD/templates/bug-report.md))
[ ] New rule ([template](https://raw.githubusercontent.com/eslint/eslint/HEAD/templates/rule-proposal.md))
[x] Changes an existing rule ([template](https://raw.githubusercontent.com/eslint/eslint/HEAD/templates/rule-change-proposal.md))
[ ] Add autofix to a rule
[ ] Add a CLI option
[ ] Add something to the core
[ ] Other, please explain:

Refs: #15703 

#### What changes did you make? (Give an overview)
Change to the existing rule **no-fallthrough**.

Added an option to allow empty cases in the switch case. 
#### Is there anything you'd like reviewers to focus on?

**What rule do you want to change?**

no-fallthrough

**Does this change cause the rule to produce more or fewer warnings?**
 yes, if allowEmptyCase option is enabled will not show any error for empty cases.
**How will the change be implemented? (New option, new default behavior, etc.)?**
  new option 
**Please provide some example code that this change will affect:**

```js
var foo = 2;
var a = function (){
    console.log('sample function');
}
switch(foo){
    case 1:
    /*
    Put a message here 
    */
    case 2: a();
}
```

**What does the rule currently do for this code?**

Throws error for the empty rule.
Error: 
``` 
Expected a 'break' statement before 'case'. ([no-fallthrough](https://eslint.org/docs/rules/no-fallthrough)) 
```

**What will the rule do after it's changed?**

```
/* eslint no-fallthrough: ["error", { "allowEmptyCase": true }] */
```
If we set allowEmptyCase to true, won't show errors for empty cases regardless of blank lines and comments inside.

### CODE REVIEW

**Review by mdjermanovic — CHANGES_REQUESTED** [review]
Thanks for the PR!

Can you also update the documentation for this rule: [docs/src/rules/no-fallthrough.md](https://github.com/eslint/eslint/blob/main/docs/src/rules/no-fallthrough.md)

File: lib/rules/no-fallthrough.js [hunk 1]
```diff
@@ -35,6 +35,19 @@ function hasFallthroughComment(caseWhichFallsThrough, subsequentCase, context, f
...
 }
 
+/**
+ * Checks whether or not a given case is empty.
+ * @param {ASTNode} caseWhichFallsThrough SwitchCase node which falls through.
+ * @returns {boolean} `true` if the case is empty.
+ */
+function isEmptyCase(caseWhichFallsThrough) {
+    if (caseWhichFallsThrough.consequent.length === 0 ||
+        (caseWhichFallsThrough.consequent.length === 1 && caseWhichFallsThrough.consequent[0].type === "EmptyStatement")) {
```
[the remark this answers is not in the document]
Comment (amareshsm) [reply lib/rules/no-fallthrough.js:45]: added in the tests.

**Comment by amareshsm** [comment]
> Thanks for the PR!
> 
> Can you also update the documentation for this rule: [docs/src/rules/no-fallthrough.md](https://github.com/eslint/eslint/blob/main/docs/src/rules/no-fallthrough.md)

updated the documentation.

File: lib/rules/no-fallthrough.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (mdjermanovic) [reply lib/rules/no-fallthrough.js:45]: Thank for clarifying!

But I'm not sure we want to allow switch cases with `;`. I checked again all posts in the original issue https://github.com/eslint/eslint/issues/15703 and couldn't find a request for this, and I don't think it's a common practice to mark empty switch cases with a `;`.

@nzakas what do you think about this?

[the remark this answers is not in the document]
Comment (nzakas) [reply lib/rules/no-fallthrough.js:45]: I agree. I don’t think an empty statement should count as a case being empty.

[the remark this answers is not in the document]
Comment (amareshsm) [reply lib/rules/no-fallthrough.js:45]: 👍 Will disallow switch cases with empty statement `;`

[the remark this answers is not in the document]
Comment (amareshsm) [reply lib/rules/no-fallthrough.js]: Changes done

File: lib/rules/no-fallthrough.js [hunk 2]
```diff
@@ -35,6 +35,22 @@ function hasFallthroughComment(caseWhichFallsThrough, subsequentCase, context, f
...
+ * Checks whether or not a given case is empty.
+ * @param {ASTNode} caseWhichFallsThrough SwitchCase node which falls through.
+ * @param {RuleContext} context A rule context which stores comments.
+ * @returns {boolean} `true` if the case is empty.
+ */
+function isEmptyCase(caseWhichFallsThrough, context) {
+    const sourceCode = context.getSourceCode();
+
+    if (caseWhichFallsThrough.consequent.length === 0 ||
+        (caseWhichFallsThrough.consequent.length === 1 && caseWhichFallsThrough.consequent[0].type === "EmptyStatement" && !sourceCode.getLastToken(caseWhichFallsThrough.consequent[0]).value === ";")) {
```
[the remark this answers is not in the document]
Comment (mdjermanovic) [reply lib/rules/no-fallthrough.js:48]: I think that EmptyStatement is causing confusion here.

EmptyStatement = lone semi

```js
switch (foo) {
    case bar: 
    ; // this is an EmptyStatement

    case baz:

    // this is not an EmptyStatement
}
```

For the new option, an empty SwitchCase is a case that doesn't have any statement inside (`consequent.length === 0`). We don't have to check for `EmptyStatement` nodes at all.

File: tests/lib/rules/no-fallthrough.js [hunk 3]
```diff
@@ -214,6 +222,59 @@ ruleTester.run("no-fallthrough", rule, {
...
+            errors: [
+                {
+                    messageId: "case",
+                    type: "SwitchCase",
+                    line: 3,
+                    column: 1
+                }
+            ]
+        },
+        {
```
Comment (mdjermanovic) [inline tests/lib/rules/no-fallthrough.js:237]: Can you also add a test case with `allowEmptyCase: false`, and a test case with an empty object (to test default value in the schema), like:

```js
{
    code: "switch(foo) { case 0:\n\ncase 1: b(); }",
    options: [{ allowEmptyCase: false }],
    errors: [
        {
            messageId: "case",
            type: "SwitchCase",
            line: 3,
            column: 1
        }
    ]
},
{
    code: "switch(foo) { case 0:\n\ncase 1: b(); }",
    options: [{}],
    errors: [
        {
            messageId: "case",
            type: "SwitchCase",
            line: 3,
            column: 1
        }
    ]
},
```

**Review by mdjermanovic — CHANGES_REQUESTED** [review]

**Comment by mdjermanovic** [comment]
@amareshsm are you still working on this? Please feel free to let us know if you need any help or more information. It would be nice to finish this, and I think it's close.

**Comment by nzakas** [comment]
@amareshsm just checking back to see if you are still working on this?

File: docs/src/rules/no-fallthrough.md [hunk 4]
```diff
@@ -163,9 +163,11 @@ Note that the last `case` statement in these examples does not cause a warning b
 
 ## Options
 
-This rule accepts a single options argument:
+This rule has an object option for exceptions:
```
[the remark this answers is not in the document]
Comment (amareshsm) [reply docs/src/rules/no-fallthrough.md]: done

File: tests/lib/rules/no-fallthrough.js (hunk 3 above)
Comment (amareshsm) [reply tests/lib/rules/no-fallthrough.js:245]: added

**Comment by amareshsm** [comment]
Changes done. PR is ready for review

File: docs/src/rules/no-fallthrough.md [hunk 5]
```diff
@@ -201,6 +203,29 @@ switch(foo) {
...
+    case 1:
+    /*
+    Put a message here 
+    */
+    case 2: doSomething();
+}
+
+```
+
 :::
```
Comment (mdjermanovic) [inline docs/src/rules/no-fallthrough.md:229]: `::: correct` before this code block and `:::` after the previous code block are missing.

File: docs/src/rules/no-fallthrough.md [hunk 6]
```diff
@@ -201,6 +203,29 @@ switch(foo) {
 }
 ```
 
+### allowEmptyCase
+
+Examples for **correct** code for the `{ "allowEmptyCase": true }` option:
```
Comment (mdjermanovic) [inline docs/src/rules/no-fallthrough.md:208]:
```suggestion
Examples of **correct** code for the `{ "allowEmptyCase": true }` option:
```

File: lib/rules/no-fallthrough.js [hunk 7]
```diff
@@ -118,12 +135,17 @@ module.exports = {
                 /*
                  * Checks whether or not there is a fallthrough comment.
                  * And reports the previous fallthrough node if that does not exist.
+                 *
+                 *
```
Comment (mdjermanovic) [inline lib/rules/no-fallthrough.js:139]:
```suggestion
```

These added empty comment lines seem unnecessary?

File: lib/rules/no-fallthrough.js [hunk 8]
```diff
@@ -118,12 +135,17 @@ module.exports = {
...
-                    });
+
+                if (fallthroughCase && (!hasFallthroughComment(fallthroughCase, node, context, fallthroughCommentPattern))) {
+                    if (!(allowEmptyCase && isEmptyCase(fallthroughCase))) {
+                        context.report({
+                            messageId: node.test ? "case" : "default",
+                            node
+                        });
+                    }
                 }
```
Comment (mdjermanovic) [inline lib/rules/no-fallthrough.js:149]:
```suggestion
                if (fallthroughCase && (!hasFallthroughComment(fallthroughCase, node, context, fallthroughCommentPattern))) {
                    context.report({
                        messageId: node.test ? "case" : "default",
                        node
                    });
                }
```

The added condition seems redundant (logic for the new option is already implemented on line 162).

**Review by mdjermanovic — CHANGES_REQUESTED** [review]

File: docs/src/rules/no-fallthrough.md [hunk 9]
```diff
@@ -169,9 +169,11 @@ Note that the last `case` statement in these examples does not cause a warning b
 
 ## Options
 
-This rule accepts a single options argument:
+This rule has an object option:
 
-* Set the `commentPattern` option to a regular expression string to change the test for intentional fallthrough comment
+* Set the `commentPattern` option to a regular expression string to change the test for intentional fallthrough comment.
+
+* Set the `allowEmptyCase` option to `true` to allow empty cases regardless of the layout.By default, this rule does not require a fallthrough comment after an empty `case` only if the empty `case` and the next `case` are on the same line or on consecutive lines.
```
[the remark this answers is not in the document]
Comment (amareshsm) [reply docs/src/rules/no-fallthrough.md]: updated.

File: lib/rules/no-fallthrough.js (hunk 8 above)
Comment (amareshsm) [reply lib/rules/no-fallthrough.js]: removed redundant logic

File: lib/rules/no-fallthrough.js (hunk 7 above)
Comment (amareshsm) [reply lib/rules/no-fallthrough.js]: removed

**Review by mdjermanovic — APPROVED** [review]
LGTM, thanks!

I'll leave this open for a few days in case someone else wants to review it before merging.

**Pull request merged**